crates/pathfinder/src/state/sync.rs

```diff
@@ -0,0 +1,331 @@
+mod l1;
+mod l2;
+
+use crate::{
+    core::{
+        ContractHash, ContractRoot, GlobalRoot, StarknetBlockHash, StarknetBlockNumber,
+        StarknetBlockTimestamp,
+    },
+    ethereum::{
+        log::StateUpdateLog,
+        state_update::{DeployedContract, StateUpdate},
+        Chain,
+    },
+    sequencer::{self, reply::Block},
+    state::{
+        calculate_contract_state_hash, state_tree::GlobalStateTree, update_contract_state,
+        CompressedContract,
+    },
+    storage::{
+        BlockId, ContractCodeTable, ContractsStateTable, ContractsTable, L1StateTable, RefsTable,
+        StarknetBlock, StarknetBlocksTable, Storage,
+    },
+};
+
+use anyhow::Context;
+use pedersen::StarkHash;
+use rusqlite::{Connection, Transaction};
+use tokio::sync::{mpsc, oneshot};
+use web3::{transports::Http, Web3};
+
+/// The sync events which are emitted by the L1 and L2 sync processes.
+#[derive(Debug)]
+enum SyncEvent {
+    L1Update(Vec<StateUpdateLog>),
+    L2Update(Block, StateUpdate),
+    L1Reorg(StarknetBlockNumber),
+    L2Reorg(StarknetBlockNumber),
+    L2NewContract(CompressedContract),
+    QueryL1Update(StarknetBlockNumber, oneshot::Sender<Option<StateUpdateLog>>),
+    QueryL2Hash(
+        StarknetBlockNumber,
+        oneshot::Sender<Option<StarknetBlockHash>>,
+    ),
+    QueryL2ContractExistance(Vec<ContractHash>, oneshot::Sender<Vec<bool>>),
```

Comment:
May also make sense to separate out the queries specifically into a separate channel.

Reply:
But wouldn't that mean that query results can become non deterministic in some scenarios?

Reply:
Not with how it operates currently -- since the reply is sent via the `oneshot` channel, it means that the query sender has to wait for the reply anyway.